p2p/net/swarm/swarm_dial.go

```diff
@@ -20,17 +95,52 @@ import (
 // This allows us to use various transport protocols, do NAT traversal/relay,
 // etc. to achive connection.
 func (s *Swarm) Dial(ctx context.Context, p peer.ID) (*Conn, error) {
-
 	if p == s.local {
 		return nil, errors.New("Attempted connection to self!")
 	}
 
-	// check if we already have an open connection first
-	cs := s.ConnectionsToPeer(p)
-	for _, c := range cs {
-		if c != nil { // dump out the first one we find
-			return c, nil
+	// this loop is here because dials take time, and we should not be dialing
+	// the same peer concurrently (silly waste). Additonally, it's structured
+	// to check s.ConnectionsToPeer(p) _first_, and _between_ attempts because we
+	// may have received an incoming connection! if so, we no longer must dial.
+	//
+	// During the dial attempts, we may be doing the dialing. if not, we wait.
+	var err error
+	var conn *Conn
+	for i := 0; i < dialAttempts; i++ {
+		// check if we already have an open connection first
+		cs := s.ConnectionsToPeer(p)
+		for _, conn = range cs {
+			if conn != nil { // dump out the first one we find. (TODO pick better)
+				return conn, nil
+			}
+		}
+
+		// check if there's an ongoing dial to this peer
+		if ok, wait := s.dsync.Lock(p); !ok {
+			select {
+			case <-wait: // wait for that dial to finish.
+				continue // and see if it worked (loop), OR we got an incoming dial.
+			case <-ctx.Done(): // or we may have to bail...
+				return nil, ctx.Err()
+			}
 		}
+
+		// ok, we have been charged to dial! let's do it.
+		conn, err = s.dial(ctx, p)
+		s.dsync.Unlock(p)
```

Comment:
It wasn't immediately clear that the conn is added to the swarm's list of conns in the dial method, so I thought this lock might've been released too early.

However, I read through and saw that swarm add conn occurs in dialConnSetup.

(nested state changes are somewhat hard to follow. It's nice to have lots of pure code (and to isolate code containing the side-effects, and to keep the side-effects as close to the surface as possible))

but the code looks correct. LGTM